Initiation for 4 log ratio-transform functions to support GPU and Array standard

songyesog2000 opened the pull request:
Hi Scikit-team,

This is Sam, I have been working on this issue as I want to help Jamie on the development also a part of practice to improve my coding.

This is just an initialization for the pull&request, as I have finish the adaption of codes for GPU and array standard api,  I will finish the test.py next two days.

There are a few points I need to consult your team about the functional behaviors:

1. typing: I used Array (a dummy class of class, implemented in array standard api), however, I notice that composition data can be a certain class/data type, if you want to add the class, please give me a hint where to add and what property/method this class should embrace.
2. closure: I noticed closure function is automatically operated within the 4 functions, however, log-ratio transformed is not necessary coupled with the closure, will it be a good idea to avoid having the closure inside these functions?
3. shape of returns: the input maybe in various shapes (vectors, matrices, singleton), should the return shape be consistent or corresponding the shape of input?

Best regards,
Sam Mingsong 

Issue reference: 
https://github.com/scikit-bio/scikit-bio/issues/2210


=============================================

Please complete the following checklist:

* [ ] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [ ] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [ ] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

qiyunzhu commented on skbio/stats/composition.py:
```diff
@@ -489,13 +525,23 @@ def clr(mat):
+    xp = aac.array_namespace(mat)
+    
+    # ensure last dimension is not singleton
+    if xp.shape(mat)[-1] == 1:
```
I wonder if this can be `mat.shape[-1]`? In my experience with NumPy, `ndarray.some_method` is usually more efficient than `np.some_function`.

songyesog2000 replied:
I think 

> I wonder if this can be `mat.shape[-1]`? In my experience with NumPy, `ndarray.some_method` is usually more efficient than `np.some_function`.

Agree that there will be some sacrifice of efficiency, however, inevitable ish.
mat can be the ndarray of any array library (say PyTorch, numpy), xp is a wrapper of correspondent library, where xp rewrite the functions but following Array API standards.

So mat.some_function will bring some risk as functional behavior maybe different, for instance, dim argument in PyTorch while axis used in numpy.

For mat.shape I think it's okay as it is likely universal for all libraries, however, the improvement on efficiency is very limited.

qiyunzhu commented on skbio/stats/composition.py:
```diff
@@ -533,12 +579,20 @@ def clr_inv(mat):
+    return xp.exp(mat + gm)
 
-def ilr(mat, basis=None, check=True):
+def ilr(mat:Array, basis:Array|None=None, /, check:bool=True)-> Array:
```
scikit-bio still supports Python 3.9, which doesn't allow `|` in type hinting. So `Array|None` needs to be `Optional[Array]` instead. Likewise, `int|Array` below needs to be `Union[int, Array]`.

mortonjt replied:
Let's not forget to remove the `/` here

qiyunzhu commented on skbio/stats/composition.py:
```diff
@@ -721,22 +779,24 @@ def alr(mat, denominator_idx=0):
-    mat = closure(mat)
-    if mat.ndim == 2:
+    xp = aac.array_namespace(mat)
+    if xp.ndim(mat) == 2:
```
I wonder if `mat.ndim` still works. I could be wrong though.

mortonjt commented:
RE closure, we don't need to call the original function, but we should have asserts place, namely making sure that there are no negative entries
https://github.com/scikit-bio/scikit-bio/blob/main/skbio/stats/composition.py#L195-L196

And making sure that the rows don't sum to zero
https://github.com/scikit-bio/scikit-bio/blob/main/skbio/stats/composition.py#L199-L200

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -590,23 +644,25 @@ def ilr(mat, basis=None, check=True):
 
 
-def ilr_inv(mat, basis=None, check=True):
+def ilr_inv(mat:Array, basis:Array|None=None, check:bool=True)-> Array:
```
let's rename this to validate_args to be consistent with other packages (i.e. PyTorch). Every function will need to have this

qiyunzhu replied:
In other functions of scikit-bio this parameter is often called `validate`. [Example 1](https://scikit.bio/docs/latest/generated/skbio.diversity.beta_diversity.html) [Example 2](https://scikit.bio/docs/latest/generated/skbio.stats.distance.DissimilarityMatrix.html). I explored around and found that there isn't a consensus outside deep learning community. For example, SciPy's [curve_fit](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.curve_fit.html) uses `check_finite`. scikit-learn doesn't have those flags but have standalone functions like [check_array](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_array.html), [check_X_y](https://scikit-learn.org/stable/modules/generated/sklearn.utils.check_X_y.html), etc.

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -721,22 +779,24 @@ def alr(mat, denominator_idx=0):
 
 
-def alr_inv(mat, denominator_idx=0):
+def alr_inv(mat:Array, denominator_idx:int|Array=0, /):
```
@qiyunzhu @songyesog2000 I don't have a super strong opinion as long as it doesn't introduce unforeseen complexity.

qiyunzhu replied:
@mortonjt  I learned about this only recently after we were asked to add type hinting... I don't have a strong opinion either. In fact, before I learned about this, I found the documentations with `*` or `/` confusing.

songyesog2000 commented:
Haven't finished yet, outstanding tasks:
1. add dependency/ test decencies into CI
2. a new unit test for arries of different array-consuming libraries.
3. efficience improvement.

tests/test_composition.py has been passed, that is, backward compatibility is ensured (i.e., works for numpy and the result is consistent)
the computational efficiency is a little compromised 

20.5s for testing the new version
<img width="1039" alt="Screenshot 2025-06-05 at 9 12 30 PM" src="https://github.com/user-attachments/assets/234e7be2-19ab-4c10-84f2-03d4ea503daa" />

19s for testing the initial version
![image](https://github.com/user-attachments/assets/c090ab37-980a-4af2-9a92-267a34b94393)

songyesog2000 commented:
updated the CI config file for the dependencies/ test dependencies, CI should be complete
added a new file for unit testing for  PyTorch (cpu, cuda) array compatibility
updated composition as basis for irv needed to be consistent on the same gpu, dtype
made clousure function compatible with Array API compat.

Outstanding work:
1. extend the unit test to JAX, CuPy, xarray
2. Redo the Documentation

songyesog2000 commented:
> @songyesog2000 There seems to be some dependency issue. Specifically, the combination of PyTorch, Python 3.13 and Intel MacOS is not available on PyPI (seemingly). If this issue cannot be resolved, a quick patch would be excluding PyTorch from the test requirement, and only test it locally and manually. [Here is how this can be done](https://github.com/scikit-bio/scikit-bio/blob/4872c6c73a0c151cada2139e4d849994e3b2121c/skbio/stats/distance/tests/test_base.py#L555).

Thank you for the heads up, will look into, also, document and unit test for Jax is almost done, will push them tonight!

songyesog2000 commented:
Updated for CI/documentation/jax.numpy

However, the testing show Jax.numpy runs extremely slow also have to reduce atol/rtol from 1e-7 to 1e-4 to pass assert_allclose.
Probably because of the default dtype (is float16?) , JIT compile, or arrays are immutable  in Jax.

qiyunzhu commented:
@songyesog2000 Error again... due to missing torch, and linting issues. For the latter, you can resolve it automatically using the [Ruff pre-commit hook](https://scikit.bio/contribute.html#style-code)

songyesog2000 commented:
Sorry, my careless, should be good to go now

songyesog2000 commented on pyproject.toml:
```diff
@@ -44,6 +44,7 @@ dependencies = [
     "biom-format >= 2.1.16",
     "statsmodels >= 0.14.0",
     "patsy >= 0.5.0",
+    "array-api-compat>=1.3,<2.0"
```
Will remove the upper bound version.

songyesog2000 commented:
Minor change is done as per suggested, thanks!

songyesog2000 commented on skbio/stats/tests/test_composition_torch_compat.py:
```diff
@@ -0,0 +1,1071 @@
+    #                 psi[i, j] = np.sqrt(s[i]/(r[i]*(r[i]+s[i])))
+    #             elif sbp[i, j] == -1:
+    #                 psi[i, j] = -np.sqrt(r[i]/(s[i]*(r[i]+s[i])))
+    #     npt.assert_allclose(psi, sbpbasis)
```
Will ditch this as we discussed in the zoom

songyesog2000 commented on skbio/stats/composition.py:
```diff
@@ -566,8 +657,14 @@ def ilr(mat, basis=None, check=True):
         Check to see if basis is orthonormal.
+    is_aitchison: bool, default False
+        Should be ignored, True when the given mat is known to be aitchison, otherwise
+        closre will be performed.
```
Shown in the attached picture, the domain of clr is S^D (the aitchison). However, the initial code also accept R_+^D, which will run closure automatically to map into S^D. This makes sense when the raw data is accounts, however, skipping this automation can speed up the calculation when the data is preprocessed or known to be aitchison.
![image](https://github.com/user-attachments/assets/0c8a7ac0-1524-404d-a40d-03ae2984454d)

mortonjt replied:
Hi @qiyunzhu , @songyesog2000 and I had an offline discussion. We should kick out this parameter, since it is completely unnecessary and will be confusing to users.

songyesog2000 commented on skbio/stats/composition.py:
```diff
@@ -532,13 +609,27 @@ def clr_inv(mat):
+    else:
+        # NOTE: for backward compatibility
+        # mat = mat - xp.sum(mat, axis=-1, keepdims=True)
+        pass
```
> @songyesog2000 I think perhaps a warning is more appropriate here. This is a valid operation that is more telling the user that their CLR array isn't a true CLR array since it doesn't sum to zero. But it is still valid, since the proportions returned from clr_inv will still be valid.

Yep, this is a better idea, will modify it.

songyesog2000 commented:
This RP regarding the compatibility with array-consuming lib is done.
The adoption to ND tensor will be in another PR.

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -634,7 +752,9 @@ def ilr_inv(mat, basis=None, check=True):
         orthonormal basis.
-    check : bool
+    axis: int, default -1
+        ignored for backward compactibility
```
I'm not crazy about introducing the axis into the function signature at this moment in time. It will make sense once we have tensor support, but for now , it will just confuse users.

I propose to have axis=-1 hard-coded within the function, so that it is not exposed to the user interface. @qiyunzhu do you agree?

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -701,10 +836,15 @@ def alr(mat, denominator_idx=0):
         specifies the first column or position.
+    axis: int, default -1
+        Should be ignored for backward compactibility, the index of dimension the
+        operation will be applied.
```
Let's move axis=-1 to the inside of the function

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -763,10 +935,13 @@ def alr_inv(mat, denominator_idx=0):
         specifies the first column or position.
+    axis: int, default -1
+        Should be ignored for the backward compactivity, the index of dimension
+        that the operation will be applied on.
```
Same here, move axis to inside of function

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -590,23 +676,55 @@ def ilr(mat, basis=None, check=True):
-    return clr(mat) @ basis.T
+    # switch the wanted axis to the last one
+    if axis != -1:
+        return xp.permute_dims(mat @ basis.T, switch_tuple)
```
Remove since axis=-1

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -721,22 +861,54 @@ def alr(mat, denominator_idx=0):
+
+    # recover the permutation
+    if axis != -1:
+        mat = mat.permute_dims(switch_tuple)
```
remove since axis=-1

mortonjt reviewed (commented):
Almost there. Let's have the following done before merging

1. Update the changelog
2. Remove all axis arguments in this pull request
3. Remove all switch_tuple variables

We can always revisit permute dims and axis arguments in the next pull request when dealing with N-D tensors

songyesog2000 commented:
modified as suggested, ruff check passed.
Did not involve with changelog, looked at CHANGELOG.md feeling not the thing I should touch.

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -590,23 +668,48 @@ def ilr(mat, basis=None, check=True):
+        xp = aac.array_namespace(mat)
+    if validate:
+        # assert from closure() while it is removed in latest version
+        if xp.any(mat < 0):
```
isn't this redundant? This check is contained inside of `_composition_check`

songyesog2000 replied:
Yep, this is redundant, actually, not for the _composition_check, but for clr(mat,validate).
mat will be checked again in clr if validate is true.

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -489,13 +527,36 @@ def clr(mat):
+    if validate:
+        # assert from closure() while it is removed in latest version
+        if xp.any(mat < 0):
+            raise ValueError("Cannot have negative proportions")
```
Same here, isn't his redundant since this check is contained inside of `_composition_check`?

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -721,22 +836,46 @@ def alr(mat, denominator_idx=0):
+                f"dimension D{mat.ndim + axis} of the input matrix is singleton"
+            )
+        if xp.any(mat <= 0):
+            raise ValueError("Input matrix must be positive")
```
Shouldn't we use `_composition_check` here instead for consistency?

mortonjt commented on skbio/stats/composition.py:
```diff
@@ -782,26 +921,52 @@ def alr_inv(mat, denominator_idx=0):
+        #     "the input matrix has more than 2 dimensions, \
+        #               high dimensional alr is new feature",
+        #     UserWarning,
+        # )
```
Remove these comments

mortonjt reviewed (commented):
Hi @songyesog2000 , last round of comments. Don't forget to update the changelog -- we have a habit of not merging PRs until the changelog is updates

https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md

See here for more details why this is important https://scikit.bio/contribute.html#changelog

songyesog2000 commented on skbio/stats/composition.py:
```diff
@@ -782,26 +921,52 @@ def alr_inv(mat, denominator_idx=0):
+
+        mat = np.asarray(mat)
+        xp = aac.array_namespace(mat)
+
```
@mortonjt how do you think about adding validate in alr_inv and other inv functions. Initially there is not such setting.

mortonjt replied:
lets not - not clear what we are validating

songyesog2000 commented:
Sorry... I don't know how this PR work, but I committed  the new one resolved the comments.
My bad, should not mark the comment resolved before pushing the commit.

songyesog2000 commented on skbio/stats/composition.py:
```diff
@@ -157,6 +157,30 @@
+
+Array = object
+
+def supports_array_api(obj):
```
> @songyesog2000 Quick question: What is this `supports_array_api` function for? It isn't used anywhere in the codebase. Can this be removed? (Just a question not a request. I will remove it if it can.)

@qiyunzhu  it is a function to check whether the library of the object supports array_api (standards) , some package, e.g. numpy, have already adopted array_api under the __array_namespace__ . Please remove it, it is redundant  as it is implicitly integrated into the wrapper  array_api_compat already.

My initial intention is to detect the case when the input is of list type to avoid using "try", e.g. code line 841-850. However, I just realize using  `supports_array_api` will also skip some libraries that  array_api_compat supports, so discarded.

qiyunzhu commented on skbio/stats/composition.py:
```diff
@@ -489,13 +527,36 @@ def clr(mat):
+        _composition_check(mat)
+
+    original_shape = mat.shape
+    # squeeze the singleton dimensions
```
@songyesog2000 I think that `squeeze` is part of the Python API standard ([source](https://data-apis.org/array-api/latest/API_specification/generated/array_api.squeeze.html)) and it automatically removes all singleton dimensions. So the following line can just be `mat = xp.squeeze(mat)`?

qiyunzhu commented on skbio/stats/composition.py:
```diff
@@ -191,15 +217,24 @@ def closure(mat):
+    if xp.any(mat < 0):
         raise ValueError("Cannot have negative proportions")
     if mat.ndim > 2:
         raise ValueError("Input matrix can only have two dimensions or less")
```
@mortonjt If I understand correctly, you are suggesting that the last dimension of a multidimensional array is the composition. So the axis closure is working on should become -1. The parameter signature should be `mat : array_like of shape (..., n_components)`. Is my understanding correct? In the current code, `clr` already has tensor support (axis=-1).
